### ISSUE TITLE
Investigate results of 02_45

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/almanac.httparchive.org#139]
From metric 02_45 in the [CSS analysis results sheet](https://docs.google.com/a/httparchive.org/spreadsheets/d/1uFlkuSRetjBNEhGKWpkrXo4eEIsgYelxY-qR9Pd7QpM/edit?disco=AAAADmoDVyw):

> Read as: "On desktop pages, when the class attribute is defined the median number of classes is 1"

client | p10 | p25 | p50 | p75 | p90
-- | -- | -- | -- | -- | --
desktop | 1 | 1 | 1 | 2 | 3
mobile | 1 | 1 | 1 | 2 | 3

@argyleink left this comment:

> this data doesnt look right. the p90 values should be much higher. i expected a range of 0-12
>
> look at tailwind or any OOCSS lib, and you can get anything done without at least 5 classes on an element. something smelly about these results imo.

The query is at https://github.com/HTTPArchive/almanac.httparchive.org/blob/master/sql/2019/02_CSS/02_45.sql

Need to investigate whether the results are accurate.

**Comment by rviscomi** [comment]
I made a histogram of the frequencies of each classList length:


classes | desktop | mobile
-- | -- | --
0 | 0.11% | 0.11%
1 | 64.27% | 63.39%
2 | 20.18% | 20.52%
3 | 7.44% | 7.71%
4 | 4.31% | 4.49%
5 | 1.71% | 1.75%
6 | 0.74% | 0.74%
7 | 0.38% | 0.39%
8 | 0.22% | 0.22%
9 | 0.19% | 0.19%
10 | 0.10% | 0.10%
11 | 0.06% | 0.07%
12 | 0.05% | 0.05%
13 | 0.04% | 0.04%
14 | 0.03% | 0.03%
15 | 0.02% | 0.03%

![image](https://user-images.githubusercontent.com/1120896/64366591-8a391280-cfe4-11e9-8be2-1800e570664a.png)


According to the results, 1 or 2 class names make up 80+% of all attribute values, so a p90 of 3 makes sense. 

@argyleink do you think OOCSS libraries are prolific enough to skew the distribution? In [02_10](https://docs.google.com/spreadsheets/d/1uFlkuSRetjBNEhGKWpkrXo4eEIsgYelxY-qR9Pd7QpM/edit#gid=833882504) we see very few Tailwind pages, which is probably full of false negatives from Wappalyzer, but other than Bootstrap and animate.css websites don't seem to be using CSS libraries that much.

**Comment by argyleink** [comment]
interesting.. i brought up OOCSS libs because they're a primo example of AMPLE usage of classes, like enough that the high end could/should be quite a few classes per node. I could be convinced that OOCSS libs like tachyons, tailwind, etc arent popular enough to influence the data, but even other seamingly very very popular libraries like bootstrap or strategies like BEM have more than 2 classes on them almost always. 

so sure, we could use this as a talking point, because it's counter to our assumptions. but something still feels off, like, # of classes on an element shouldnt be the most surprising result from scrubbing the entire web and comparing it to our assumptions. yet it is right now. 🤷‍♂

**Comment by rviscomi** [comment]
The total percent of all class name lengths greater than 10 is only 0.5%. We're talking about a small percent of a sample of ~1.6B class attributes though, so that's still ~8M instances of having 10+ classes. That perspective might make this more digestible. 

Fun fact: the most class names is 21,504! It only occurs once and I assume that's a parsing bug 😁 

Here's the [sheet with the full results](https://docs.google.com/spreadsheets/d/1M6cdHPfJSDVYJcaeFQPbw_NhXTIBs8qsCTG1GJpKIko/edit?usp=sharing) if you want to explore.

**Comment by argyleink** [comment]
great ideas.
that sheet is fascinating!

**Comment by una** [comment]
I think this data is super fascinating! I don't think oocss libraries are
as widely used on the web as they seem from some circles. Also I wonder if
this speaks to the majority of the web not using reusable/global styles as
frequently. The data on 0 classes being used surprised me, as I assume many
people are still styling based on base element. How would this account for
nested elements like '.list li'?

- Una

**Comment by rviscomi** [comment]
> The data on 0 classes being used surprised me, as I assume many people are still styling based on base element. How would this account for nested elements like '.list li'?

Do you mean in the CSS? This query doesn't take the selectors into account, only the `class` attributes in the HTML. So the 0 values in this case are people with empty attributes, eg `class=""`.

**Comment by obto** [comment]
This data looks pretty accurate to me.

At first I took a double take much like you all are describing. And it took me a bit to remember the majority of sites aren't "techy" like Medium, but instead use basic Wordpress themes or have a rudimentary static site like Hacker News.

After I started looking at sites like these instead, the data began making sense when I saw tons of classes like: `email`, `button`, `container`, `four columns`, `footer`, `latest_post_image clearfix`, `woocommerce single-product`
